update unit tests structure + align with BIP352 updates

josibake opened the pull request:
This involves:

* Using the new test structure: the unit test json now more closely matches the output from `getrawtransaction verbosity=2`, which should make it easy to add new test cases in the future by creating a transaction on testnet/signet and parsing the inputs directly
* Parse the pubkey from the test inputs: previously, the tests gave the pubkey directly, so none of the logic on input matching on the BIP was being tested. To parse the inputs, new functions matching the logic in the BIP are added:
  * `is_p2tr`, `is_p2wpkh`, `is_p2pkh`, `is_p2sh`
  * `get_pubkey_from_input`
* Update the hashing function: per the BIP, outpoint hashing is replaced with `hash(smallest_outpoint || A_sum)`
* Use tagged hashes: per the BIP, tagged hashing is used for the input hash, label hash, and shared secret hash

cygnet3 reviewed (commented):
Feedback mostly related to the library's api

Sosthene00 commented on src/utils.rs:
```diff
@@ -1,17 +1,170 @@
+    }
+}
+
+pub fn get_pubkey_from_input(vin: &VinData) -> Result<Option<PublicKey>, Error> {
```
Actually I'm wondering, do we really need this helper here? Wouldn't it be easier to let this function out of the sp-lib itself and handle pubkey parsing one layer above in an hypothetical sp-wallet lib? We would be able to use rust-bitcoin helper function.

josibake replied:
@Sosthene00 I go back and forth on this one :sweat_smile:. I am leaning towards keeping it (or something like it) here because public key extraction is not strictly a wallet problem: scanners/indexers need to be able to extract pubkeys from a transaction, so a function which takes `CTxIn`s (transaction inputs) and extracts public keys makes sense to me.

For the sender, it seems less relevant, since it is a wallet problem i.e. a coin selection problem. Presumably, the sender already has access the public keys without needing to "extract" them.

cygnet3 replied:
Moved to `tests` so this is resolved

josibake commented on src/utils.rs:
```diff
@@ -1,17 +1,170 @@
+        eng.input(&m.to_be_bytes());
+        LabelHash::from_engine(eng)
+    }
+    pub fn to_scalar(self) -> Scalar {
```
Slightly leaning towards having it return the `Label` directly, if that is indeed how we intend to use the scalar. Seems less foot gunny for the user.

cygnet3 replied:
Since this is in the `utils` feature I indeed think it is fine either way

josibake commented on src/utils.rs:
```diff
@@ -1,17 +1,170 @@
+pub struct VinData {
+    pub script_sig: Vec<u8>,
+    pub txinwitness: Vec<Vec<u8>>,
+    pub prevout: Vec<u8>,
```
renamed to `script_pub_key`

josibake commented:
Updated:

* Renamed `prevout` -> `script_pub_key`
* Fixed a compile error in tagged hashes commit (each commit should now compile and pass the tests)
* Added fixup commit from @Sosthene00

josibake commented:
Updated `tests/common/utils.rs`:

* Bump secp256k1 to `0.28.1`  and removed the `bitcoin-hashes-std` feature flag
* hash the message directly using `bitcoin_hashes` and create a `Message` using `from_digest`

josibake commented:
Updated:

* Move-only - moves the `get_input_from_pubkey` and related helper functions `utils.rs` -> `tests/common/utils.rs`

Sosthene00 commented:
LGTM
